pennylane_lightning/lightning_qubit.py

```diff
@@ -354,40 +408,61 @@ def apply_lightning(self, state, operations):
-            # use_csingle
-            sim = StateVectorC64(state_vector)
+        dim = 1 << self.num_wires
+        batch_size_state = self._get_batch_size(state, (dim,) * self.num_wires, dim)
+
+        if batch_size_state:
+            # We have already a batched state nothing need to be done.
+            batched_states = state
         else:
-            # self.C_DTYPE is np.complex128 by default
-            sim = StateVectorC128(state_vector)
+            # If the state is not batched, we need to properly batch it,
+            # even if there is no parameter broadcasting in play.
+            batched_states = (
+                np.repeat(state[np.newaxis, ...], batch_size_op, axis=0)
+                if batch_size_op
+                else state[np.newaxis, ...]
+            )
 
-        # Skip over identity operations instead of performing
-        # matrix multiplication with the identity.
-        skipped_ops = ["Identity"]
+        for batch_ind, state in enumerate(batched_states):
+            state_vector = np.ravel(state)
 
-        for o in operations:
-            if o.base_name in skipped_ops:
-                continue
-            name = o.name.split(".")[0]  # The split is because inverse gates have .inv appended
-            method = getattr(sim, name, None)
-
-            wires = self.wires.indices(o.wires)
-            if method is None:
-                # Inverse can be set to False since qml.matrix(o) is already in inverted form
-                method = getattr(sim, "applyMatrix")
-                try:
-                    method(qml.matrix(o), wires, False)
-                except AttributeError:  # pragma: no cover
-                    # To support older versions of PL
-                    method(o.matrix, wires, False)
+            if self.use_csingle:
+                sim = StateVectorC64(state_vector)  # use_csingle
             else:
-                inv = o.inverse
-                param = o.parameters
-                method(wires, inv, param)
-
-        return np.reshape(state_vector, state.shape)
+                # self.C_DTYPE is np.complex128 by default
+                sim = StateVectorC128(state_vector)
+
+            # Skip over identity operations instead of performing
+            # matrix multiplication with the identity.
+            skipped_ops = ["Identity"]
+            for o in operations:
+                if o.base_name in skipped_ops:
+                    continue
+                name = o.name.split(".")[0]  # The split is because inverse gates have .inv appended
+                method = getattr(sim, name, None)
+
+                wires = self.wires.indices(o.wires)
+                if method is None:
+                    # Inverse can be set to False since qml.matrix(o) is already in inverted form
+                    method = getattr(sim, "applyMatrix")
+
+                    try:
+                        _matrix = qml.matrix(o)[batch_ind] if o.batch_size else qml.matrix(o)
```

Comment:
This means that we compute the batched matrix in each iteration, but then only use one matrix of the batch, right? Can we improve on that, e.g. by first iterating over the operations and then over the batch dimension? This way, for each operation we may compute the batch of matrices, and cache them for one loop over the batch dimension.

Reply:
As a matter of fact, this wouldn't improve the calculation. Lightning needs to initialize a StateVector instance (a simulator) for each batch, moving it to an inner loop would increase the number of times the python and C++ layers would need to communicate data. It is more performant to instantiate the simulator one time and go through the operations, even if you need to call for the same matrix a number of times. We would expect that, for lightning, in normal operating conditions the chance of having a large state vector to manipulate is larger than having:  1) a large number of parameters to broadcast and 2) to build several matrices.

Reply:
Ohhh, okay, I see.
I suppose for the usual use case of lightning, the computation time of `qml.matrix(o)` indeed is negligible in many cases. Would there be any use in precomputing matrices nonetheless, e.g. for rare cases of very large (~Hilbert space dimension size) operation matrices? :thinking: Or is this a prohibitive regime anyways?
Sorry in case this is a stupid question, I'm not too used to the lightning mindset :D

Reply:
Ah, thank you a lot for the clarification, I somehow missed that this happens in the `method is None` branch when considering the repeated `qml.matrix` call! :grimacing: Sounds good to me!